File: packages/typescript/src/lint/engine.ts
/** Lint engine: iterate workflows × rules and collect violations. */

import { existsSync, statSync } from "node:fs";
import { resolve } from "node:path";
import type { App } from "../app.js";
import type { Lockfile } from "../pin/lockfile.js";
import { readLockfile } from "../pin/lockfile.js";
import type { LintConfig } from "./config.js";
import { ALL_RULES, makeRuleContext } from "./rules/index.js";
import type { Violation } from "./violation.js";

/**
 * Run all enabled lint rules against every workflow in the app.
 *
 * Actions (`action.yml`) are not currently linted — v1 scope is
 * workflows only.
 *
 * A rule that throws is caught and skipped; other rules continue. A
 * warning is written to stderr naming the failing rule.
 */
export function runLint(app: App, config: LintConfig): Violation[] {
  // Load the lockfile if available so rules can check pin status.
  let lockfile: Lockfile | null = null;
  if (app.lockfilePath !== null) {
    const full = resolve(app.root, app.lockfilePath);
    if (existsSync(full) && statSync(full).isFile()) {
      lockfile = readLockfile(full);
    }
  }

  const violations: Violation[] = [];

  for (const { item, relPath } of app._items) {
    if (item._kind !== "workflow") continue;
    const workflowKey = stem(relPath);
    const ctx = makeRuleContext(workflowKey, config, lockfile);

    for (const ruleFn of ALL_RULES) {
      const ruleId = ruleFn.meta.id;
      if (config.disable.has(ruleId)) continue;
      try {
        for (const violation of ruleFn(item, ctx)) {
          violations.push(violation);
        }
      } catch (err) {
        process.stderr.write(
          `warning: rule '${ruleId}' crashed on workflow '${workflowKey}' ` +
            `— skipped (${(err as Error).name}: ${(err as Error).message})\n`,
        );
        continue;
      }
    }
  }

  return violations;
}

function stem(path: string): string {
  const base = path.replace(/^.*[\\/]/, "");
  const dot = base.lastIndexOf(".");
  return dot > 0 ? base.slice(0, dot) : base;
}

File: packages/python/src/ghagen/lint/engine.py
"""Lint engine: iterate workflows × rules and collect violations."""

from __future__ import annotations

import sys

from ghagen.app import App
from ghagen.lint.config import LintConfig
from ghagen.lint.rules import ALL_RULES
from ghagen.lint.rules._base import RuleContext
from ghagen.lint.violation import Violation
from ghagen.models.workflow import Workflow


def run_lint(app: App, config: LintConfig) -> list[Violation]:
    """Run all enabled lint rules against every Workflow in the app.

    Actions (``action.yml``) are not currently linted — v1 scope is
    workflows only.

    A rule that raises an exception is caught and skipped; other rules
    continue. A warning is printed to stderr naming the failing rule.

    Args:
        app: The ghagen :class:`~ghagen.App` with registered workflows.
        config: The :class:`~ghagen.lint.config.LintConfig` to apply.

    Returns:
        A list of :class:`~ghagen.lint.violation.Violation`s in the order
        they were produced.
    """
    # Load lockfile if available so rules can check pin status.
    lockfile = None
    if app.lockfile_path is not None:
        lockfile_full = app.root / app.lockfile_path
        if lockfile_full.is_file():
            from ghagen.pin.lockfile import read_lockfile

            lockfile = read_lockfile(lockfile_full)

    violations: list[Violation] = []

    for item, path in app._items:
        if not isinstance(item, Workflow):
            continue

        workflow_key = path.stem  # e.g. "ci" from "ci.yml"
        ctx = RuleContext(workflow_key=workflow_key, config=config, lockfile=lockfile)

        for rule_fn in ALL_RULES:
            rule_id = rule_fn.meta.id
            if rule_id in config.disable:
                continue
            try:
                for violation in rule_fn(item, ctx):
                    violations.append(violation)
            except Exception as exc:  # noqa: BLE001 — rule isolation is the point
                print(
                    f"warning: rule '{rule_id}' crashed on workflow "
                    f"'{workflow_key}' — skipped ({type(exc).__name__}: {exc})",
                    file=sys.stderr,
                )
                continue

    return violations
